### DISCUSSION TITLE
Artemis MQ connection

### DISCUSSION THREAD

**Discussion by Renboo** [jmstoolbox/jmstoolbox#180] [Q&A]
Hello!

I try to use JmsToolbox with Artemis MQ. And it is not work.

When i use it from my app i use connection string with parameters:
1. 2 addresses
2. sslEnabled
3. trustStorePath
4. trustStorePassword
5. trustStoreType
6. keyStorePath
7. keyStorePassword
8. keyStoreType
And it works great.

Which Queue Manager shoud i use in JmsToolbox ? Apache Active MQ?
Why two other Queue Managers do not have any keyStore properties? Could i add it?

**Comment by titou10titou10** [comment]
> I try to use JmsToolbox with Artemis MQ. And it is not work.

What is the problem? what are the symptoms? where are the logs? etc...
What version of Artemis are you using?
Did you follow instructions [here](https://github.com/jmstoolbox/jmstoolbox/wiki/2.6-Setup-for-Apache-Artemis) ?
If the server is 2.x, the Queue Manager to use is "Apache Active MQ Artemis v2.x"
Your "app" (?) uses keyStorePath/keyStorePassword/keyStoreType ...is your server configured to validate a client certificate to connect?

**Comment by Renboo** [comment]
I have Artemis MQ 2.19.1 and it is fully worked with MTLS. Thats why i use keystore and truststore.
If i try to use Queue Manager "Apache Active MQ Artemis v2.x" i do not have any props for keyStore. And other requared option is StoreType. We use PCKS12 but by default java uses JKS.

Main idea to ask you to add more possible properties for connection. Is it possible?

Just for info: i use artemis-jakarta-client:2.33.0 in my test app.

Of cause i coud give you logs, but now it doesn't make sense because i could not use keyStore in "Apache Active MQ Artemis v2.x", am i right?

**Comment by titou10titou10** [comment]
OK
You still didn't say what the problem is...
What error message do you receive when you use the "Apache Active MQ Artemis v2.x" QM, set the "trust*" proporties in the additions of other properties etc. What message do you receive? what is written in the logs?

keyStorePath/keyStorePassword/keyStoreType etc.  properties need to be set when the server validates the validity of a client with a client certificate that it can verify, it has to be configured for that. This is quite a very uncommon setup... Is that your case? if not you do not need to set those properties

I just released v6.7.0 and I added the "truststoreType" property

**Comment by Renboo** [comment]
Thank you a lot for new release!

I have no idea why but today in many companies MTLS access is used as default. I think that it is very security. It goes together with autocertification that mean new certs will be issued after new deploy of app. And you do not need login and password that you should change time by time.

I have tried any combinations of properties but any time i received error:

```
!MESSAGE Connect unsuccessful
!STACK 0
ActiveMQNotConnectedException[errorType=NOT_CONNECTED message=AMQ219007: Cannot connect to server(s). Tried with all available servers.]
	at org.apache.activemq.artemis.core.client.impl.ServerLocatorImpl.createSessionFactory(ServerLocatorImpl.java:729)
	at org.apache.activemq.artemis.jms.client.ActiveMQConnectionFactory.createConnectionInternal(ActiveMQConnectionFactory.java:910)
	...
```
Looks like JMSToolBox uses different port or something else. But in Servers i have indicated 61616 that work for my test app.

Should i add keyStore* props in z_ExtraNettyProperties prop? With wich delimiter: ";" or "&"?

Is it possible to log full connection String in logs of JMSToolBox? I could be wrong but i see it for "Apache Active MQ" Queue manager in logs.

**Comment by titou10titou10** [comment]
I've added the keystore* properties for the next version of JTB
You can test it from the latest github build here:

https://github.com/jmstoolbox/jmstoolbox/actions/runs/10857826351
https://github.com/jmstoolbox/jmstoolbox/actions/runs/10857826351/artifacts/1932194501

**Comment by Renboo** [comment]
Thank you a lot for help!

Got your dev branch, build it, tried it and get same error.

I tried another way: repeat your code in my test app because i used connection string but you use props and you went other way to create connection factory in ActiveMQArtemis2QManager class.
All was ok, but i find that you made some mistakes with copies of string in lines 185-210 in ActiveMQArtemis2QManager class. You could find duplicate values for connectionParams map. 
You could remove TransportConstants.KEYSTORE_ALIAS_PROP_NAME. It is not needed now.

Additional help info:
All times error was in line:
`Connection jmsConnection = cfJMS.createConnection(sessionDef.getActiveUserid(), sessionDef.getActivePassword());`

Anyway i will try to make some fixes in ActiveMQArtemis2QManager and will try it.

**Comment by Renboo** [comment]
After fixes JMSToolbox connected successfully. Hooray! Certificate connecton works!

I will test other functions on next week.

**Comment by Renboo** [comment]
I found problem which is related to the definition of the queue type.
Should i start other discussion or issue or we could continue here?

**Comment by titou10titou10** [comment]
thanks for pointing out the problems
yes open another discussion
